### PR TITLE
Anchor propagation

### PR DESCRIPTION
**Pull request by cmyr**
This implements basic anchor propagation in `glyphs-reader`. Propagation happens before the conversion to IR. Handling this here lets us more closely match the behaviour of glyphs.app.

This work is based on a python rewrite of the original objc; that work (by @anthrotype) is available in [a google doc](https://docs.google.com/document/d/1bXF8usWtsInYFUYIMvq3ds4jjPdbf2uDykewYyxQ-wU/edit?resourcekey=0-EqPApxblj5YTtDOG-1bZfA).

This code is now fairly well covered by tests, and I am reasonably confident that it replicates the behaviour of the original code in at least all the obvious cases.

~The original code is tricky and handles a bunch of edge cases. Some of that is included in this, and some of that is not. In particular, this does not handle the "*origin" anchor (see #769) and it does not handle the possible presence of the "anchor" field on a component, which we do not currently parse from the source.~

~This includes some tests for the most common cases, as well as some nice infra for adding additional tests in the future. This code is branchy and complex, and there are doubtless a bunch we aren't hitting. In particular there is special logic to swap top/bottom, left/right, and entry/exit in cases where a component has a rotation applied. I'm not even really sure what the use case is here, and thus what a good test case would be, but perhaps it's for something like using a 'u' as an 'n' by just rotating a component?~

### CODE REVIEW

File: glyphs-reader/src/propagate_anchors.rs [hunk 1]
```diff
@@ -0,0 +1,883 @@
...
+    // to make this work, we write the anchors to a separate data structure, and
+    // then only update the actual glyphs after we've done all the work.
+    let mut all_anchors = HashMap::new();
+    for name in todo {
+        // temporarily remove the glyph so we can mess with it without borrowing all glyphs
+        let glyph = glyphs.get(&name).unwrap();
+        // we work with indices here to also get around borrowck, because
+        // if we borrow the layers in the loop we can't set the anchors in the loop body
+        for layer in &glyph.layers {
+            let anchors = anchors_traversing_components(
```
Comment (rsheeter) [inline glyphs-reader/src/propagate_anchors.rs:46]: Optional, might just be me, I'd have grasped something like `reachable_anchors` faster

File: glyphs-reader/src/propagate_anchors.rs [hunk 2]
```diff
@@ -0,0 +1,883 @@
...
+    );
+}
+
+/// Return the anchors for this glyph, including anchors from components
+fn anchors_traversing_components<'a>(
+    glyph: &'a Glyph,
+    layer: &'a Layer,
+    glyphs: &BTreeMap<SmolStr, Glyph>,
+    // map of glyph -> anchors, per-layer, updated as we do each glyph
+    visited_anchors: &HashMap<SmolStr, Vec<Vec<Anchor>>>,
```
Comment (rsheeter) [inline glyphs-reader/src/propagate_anchors.rs:90]: `visited` in a graph traversal led me to expect a fairly specific pattern of use (if already visited, return that answer, otherwise compute and record) that this doesn't follow. I wonder if it's worth renaming to just `anchors`

File: glyphs-reader/src/propagate_anchors.rs [hunk 3]
```diff
@@ -0,0 +1,883 @@
...
+    let prev_names: Vec<_> = layer.anchors.iter().map(|a| &a.name).collect();
+    let new_names: Vec<_> = anchors.iter().map(|a| &a.name).collect();
+    log::info!(
+        "propagated anchors for ('{}': {prev_names:?} -> {new_names:?}",
+        glyph.name,
+    );
+}
+
+/// Return the anchors for this glyph, including anchors from components
+fn anchors_traversing_components<'a>(
```
Comment (rsheeter) [inline glyphs-reader/src/propagate_anchors.rs:85]: This seems to do more than the name or comment suggests, for example I see application of adjustments like origin. Suggest updating at least the comment.

File: glyphs-reader/src/propagate_anchors.rs [hunk 4]
```diff
@@ -0,0 +1,883 @@
...
+    base_glyph_counts: &mut HashMap<(SmolStr, &'a str), usize>,
+) -> Vec<Anchor> {
+    if layer.anchors.is_empty() && layer.components().count() == 0 {
+        return Vec::new();
+    }
+
+    // if this is a mark and it has anchors, just return them.
+    if !layer.anchors.is_empty() && glyph.category == Some(Category::Mark) {
+        let mut anchors = layer.anchors.clone();
+        // if there is an '*origin' anchor, other positions are adjusted by its pos:
```
Comment (rsheeter) [inline glyphs-reader/src/propagate_anchors.rs:101]: Where do we assure ourselves that there aren't multiple origin anchors? - either comment or return an error

File: glyphs-reader/src/propagate_anchors.rs [hunk 5]
```diff
@@ -0,0 +1,883 @@
...
+    // map of glyph -> anchors, per-layer, updated as we do each glyph
+    visited_anchors: &HashMap<SmolStr, Vec<Vec<Anchor>>>,
+    // each (glyph, layer) writes its number of base glyphs into this map during traversal
+    base_glyph_counts: &mut HashMap<(SmolStr, &'a str), usize>,
+) -> Vec<Anchor> {
+    if layer.anchors.is_empty() && layer.components().count() == 0 {
+        return Vec::new();
+    }
+
+    // if this is a mark and it has anchors, just return them.
```
Comment (rsheeter) [inline glyphs-reader/src/propagate_anchors.rs:98]: "just return them" is a bit misleading given that we are adjusting them.

File: glyphs-reader/src/propagate_anchors.rs [hunk 6]
```diff
@@ -0,0 +1,883 @@
...
+        if let Some(origin) = anchors
+            .iter()
+            .position(Anchor::is_origin)
+            .map(|idx| anchors.remove(idx))
+        {
+            anchors
+                .iter_mut()
+                .for_each(|a| a.pos -= origin.pos.to_vec2())
+        }
+        return anchors;
```
Comment (rsheeter) [inline glyphs-reader/src/propagate_anchors.rs:111]: Suggest extracting this into it's own function

File: glyphs-reader/src/propagate_anchors.rs [hunk 7]
```diff
@@ -0,0 +1,883 @@
...
+    }
+
+    let is_ligature = glyph.sub_category == Some(Subcategory::Ligature);
+    let mut has_underscore = layer
```
Comment (rsheeter) [inline glyphs-reader/src/propagate_anchors.rs:115]: A name that indicates what has_underscore means would help.

There seems to be quite a lot of anchor name comparison with strings, many of which are repeated ("_", "entry", "exit", etc appear several times). I wonder if we might either make helper functions like `is_exit` or `is_whatever_starts_with_underscore_means` or even something like an AnchorType enum that can be build from the name.

File: glyphs-reader/src/propagate_anchors.rs [hunk 8]
```diff
@@ -0,0 +1,883 @@
...
+        .anchors
+        .iter()
+        .any(|anchor| anchor.name.starts_with('_'));
+
+    let mut number_of_base_glyphs = 0usize;
+    // we use an index map so we get the same ordering behaviour as python
```
Comment (rsheeter) [inline glyphs-reader/src/propagate_anchors.rs:121]: point to the python source here and wherever else applicable?

File: glyphs-reader/src/propagate_anchors.rs (hunk 3 above)
Comment (rsheeter) [reply glyphs-reader/src/propagate_anchors.rs:85]: It's also quite long and hard to follow, I wonder if it could be broken down into smaller functions such that this and they end up simpler?

File: glyphs-reader/src/propagate_anchors.rs [hunk 9]
```diff
@@ -0,0 +1,883 @@
...
+    const ROUND_TO: f64 = 1e6;
+    let mut pos = (transform * anchor.pos).to_vec2();
+    pos *= ROUND_TO;
+    pos = pos.round();
+    pos /= ROUND_TO;
+    anchor.pos = pos.to_point();
+}
+
+fn maybe_rename_component_anchor(comp_name: SmolStr, anchors: &mut [Anchor]) {
+    // e.g, go from 'top_1' to 'top'
```
Comment (rsheeter) [inline glyphs-reader/src/propagate_anchors.rs:276]: Just to confirm, we definitely want the _first_ `_` or assume there is only one of them?

File: glyphs-reader/src/propagate_anchors.rs [hunk 10]
```diff
@@ -0,0 +1,883 @@
...
+
+        let new_glyph = glyphs.get("commaturnedabovecomb").unwrap();
+        assert_eq!(
+            new_glyph.layers[0].anchors,
+            [("_top", (300., 502.)), ("mytop", (312., 810.)),]
+        );
+    }
+
+    #[test]
+    fn affine_scale() {
```
Comment (rsheeter) [inline glyphs-reader/src/propagate_anchors.rs:829]: This looks like a bunch of independent tests that don't need to be one big test?

**Review by rsheeter — APPROVED** [review]

File: glyphs-reader/src/propagate_anchors.rs [hunk 11]
```diff
@@ -0,0 +1,883 @@
...
+            assert_eq!(layers.len(), glyph.layers.len());
+            for (i, layer_anchors) in layers.into_iter().enumerate() {
+                glyph.layers[i].anchors = layer_anchors;
+            }
+        }
+    }
+}
+
+fn debug_new_anchors(anchors: &[Anchor], glyph: &Glyph, layer: &Layer) {
+    if !log::log_enabled!(log::Level::Info) {
```
[the remark this answers is not in the document]
Comment (cmyr) [reply glyphs-reader/src/propagate_anchors.rs:73]: oh yea i just confused trace & info

File: glyphs-reader/src/propagate_anchors.rs (hunk 1 above)
Comment (cmyr) [reply glyphs-reader/src/propagate_anchors.rs:46]: This name was chosen to match the source impl, happy to change...

File: glyphs-reader/src/propagate_anchors.rs (hunk 4 above)
Comment (cmyr) [reply glyphs-reader/src/propagate_anchors.rs:101]: anchor names are unique.

File: glyphs-reader/src/propagate_anchors.rs (hunk 6 above)
Comment (cmyr) [reply glyphs-reader/src/propagate_anchors.rs:111]: nice we can also reuse that in the other place we handle "*origin", below.

File: glyphs-reader/src/propagate_anchors.rs (hunk 8 above)
Comment (cmyr) [reply glyphs-reader/src/propagate_anchors.rs:121]: that code isn't public, and there also isn't much to point to in this case except for the fact that python dictionaries track insertion order.

File: glyphs-reader/src/propagate_anchors.rs (hunk 9 above)
Comment (cmyr) [reply glyphs-reader/src/propagate_anchors.rs:276]: the original code is assuming only one.

File: glyphs-reader/src/propagate_anchors.rs (hunk 10 above)
Comment (cmyr) [reply glyphs-reader/src/propagate_anchors.rs:829]: just me sanity checking the behaviour of this fn, doesn't feel like much of an improvement to break this into five two line tests..

File: glyphs-reader/src/propagate_anchors.rs (hunk 7 above)
Comment (cmyr) [reply glyphs-reader/src/propagate_anchors.rs:114]: - this is the original name, and the original logic, and the original code doesn't really document its intention.
- I tried the helpers, but unfortunately half the places where we're checking these things we're just working with strings, so we can't really add a method without getting fancy.
- the enum thing is nice and we do that once we get to IR, but I think down here we just have to do our best to match the glyphs code without trying to be more systematic than it is.

File: glyphs-reader/src/propagate_anchors.rs (hunk 3 above)
Comment (cmyr) [reply glyphs-reader/src/propagate_anchors.rs:92]: I did a lot of that, but most of the rest of it is pretty entwined, and the more refactoring we do the harder it is for us to compare back to the original (particularly in the future, where the reference impl might change)

File: glyphs-reader/src/propagate_anchors.rs (hunk 7 above)
Comment (rsheeter) [reply glyphs-reader/src/propagate_anchors.rs:114]: Even if we just made a function it would be an improvement over repeating the same string comparison everywhere.

If we have questions about the intent we can ask Georg, perhaps we should assemble a list of questions?

Comment (cmyr) [reply glyphs-reader/src/propagate_anchors.rs:114]: okay, added 'is_exit' and 'is_entry' helpers.


Ultimately here it feels like the implementation is the spec; there are a lot of heuristics and squishy bits here that seem like they're probably the result of trying to get some specific pattern to work. It would be nice if there was something in the glyphs docs that detailed the naming convention for anchors, but for this code in particular I don't have questions, exactly; I'm just taking it at face value and trying to match the behaviour. This is a very literal port.

File: glyphs-reader/src/font.rs [hunk 12]
```diff
@@ -2051,6 +2073,14 @@ fn preprocess_unparsed_plist(s: &str) -> Cow<str> {
 
 impl Font {
     pub fn load(glyphs_file: &path::Path) -> Result<Font, Error> {
+        let mut font = Self::load_impl(glyphs_file)?;
+        font.propagate_all_anchors();
```
Comment (anthrotype) [inline glyphs-reader/src/font.rs:2077]: I feel like this `propagate_all_anchors ` should be optional somehow, or at least there should be a method to _only_ load the .glyphs file without the anchor propagation.. It's not our current use case in fontc, but it's conceivable that one may wish to read a .glyphs file, change some bit that's unrelated to anchors, and write it back to a file to do something else.

**Comment by anthrotype** [comment]
Thanks Colin for implementing this. I know it must not have been fun 👍

File: glyphs-reader/src/propagate_anchors.rs (hunk 7 above)
Comment (rsheeter) [reply glyphs-reader/src/propagate_anchors.rs:114]: If we don't understand the naming conventions why don't we ask the Glyphs folks? - decent chance they'd be up for explaining.
